xgi/core/views.py

```diff
@@ -790,3 +792,375 @@ def maximal(self, strict=False):
+                    for key in self
+                ]
+            else:
+                raise XGIError(f"Unrecognized dtype {dtype}")
+
+        if e not in self:
+            raise IDNotFound(f'ID "{e}" not in this view')
+
+        return set(self._id_dict[e]["in"].union(self._id_dict[e]["out"]))
+
+    def head(self, e=None, dtype=list):
+        """Get the node ids that are in the head of a directed edge.
+
+        Parameters
+        ----------
+        e : hashable, optional
+            Edge ID. By default, None.
+        dtype : {list, dict}, optional
+            Specify the type of the return value.
+            By default, list.
+
+        Returns
+        -------
+        list (if dtype is list, default)
+            Head members.
+        dict (if dtype is dict)
+            Head members.
+        set (if e is not None)
+            Members of the head of edge e.
+
+        Raises
+        ------
+        TypeError
+            If `e` is not None or a hashable
+        XGIError
+            If `dtype` is not dict or list
+        IDNotFound
+            If `e` does not exist in the hypergraph
+
+        """
+        if e is None:
+            if dtype is dict:
+                return {key: self._id_dict[key]["out"].copy() for key in self}
+            elif dtype is list:
+                return [self._id_dict[key]["out"].copy() for key in self]
+            else:
+                raise XGIError(f"Unrecognized dtype {dtype}")
+
+        if e not in self:
+            raise IDNotFound(f'ID "{e}" not in this view')
+
+        return self._id_dict[e]["out"].copy()
+
+    def tail(self, e=None, dtype=list):
+        """Get the node ids that are in the tail of a directed edge.
+
+        Parameters
+        ----------
+        e : hashable, optional
+            Edge ID. By default, None.
+        dtype : {list, dict}, optional
+            Specify the type of the return value.
+            By default, list.
+
+        Returns
+        -------
+        list (if dtype is list, default)
+            Tail members.
+        dict (if dtype is dict)
+            Tail members.
+        set (if e is not None)
+            Tail members of edge e.
+
+        Raises
+        ------
+        TypeError
+            If `e` is not None or a hashable
+        XGIError
+            If `dtype` is not dict or list
+        IDNotFound
+            If `e` does not exist in the hypergraph
+
+        """
+        if e is None:
+            if dtype is dict:
+                return {key: self._id_dict[key]["in"].copy() for key in self}
+            elif dtype is list:
+                return [self._id_dict[key]["in"].copy() for key in self]
+            else:
+                raise XGIError(f"Unrecognized dtype {dtype}")
+
+        if e not in self:
+            raise IDNotFound(f'ID "{e}" not in this view')
+
+        return self._id_dict[e]["in"].copy()
+
+    def sources(self, e=None, dtype=list):
```

Comment:
Did we also have these duplicate methods with different names?

Reply:
Yes, this was a request when I originally implemented dihypergraphs for explainability. Would you like me to remove them?